src/rpc/notifications.cpp

```diff
@@ -0,0 +1,381 @@
+
+    CCriticalSection m_cs_notification_queue;
+
+    const std::string typeToString(NotificationType type) const
+    {
+        switch (type) {
+        case NotificationType::Block:
+            return MSG_HASHBLOCK;
+            break;
+        case NotificationType::Tx:
+            return MSG_HASHTX;
+            break;
+        default:
+            return "unknown";
+        }
+    }
+
+    NotificationType stringToType(const std::string& strType) const
+    {
+        if (strType == MSG_HASHBLOCK)
+            return NotificationType::Block;
+        else if (strType == MSG_HASHTX)
+            return NotificationType::Tx;
+        else
+            return NotificationType::Unknown;
+    }
+
+    // populates a json object with all notifications in the queue
+    // returns a range to allow removing the elements from the queue
+    // after successfull transmitting
+    queueRange_t weakDequeueNotifications(UniValue& result)
+    {
+        size_t firstElement = 0;
+        size_t elementCount = 0;
+
+        LOCK(m_cs_notification_queue);
+        for (const NotificationEntry& entry : m_queue) {
+            UniValue obj = UniValue(UniValue::VOBJ);
+            obj.pushKV("type", typeToString(entry.m_type));
+            obj.pushKV("seq", entry.m_sequence_number);
+            obj.pushKV("obj", entry.m_notification);
+            result.push_back(obj);
+            elementCount++;
+        }
+        return std::make_pair(firstElement, elementCount);
+    }
+
+    // removes notifications in the given range from the queue
+    void eraseRangeFromQueue(const queueRange_t range)
+    {
+        LOCK(m_cs_notification_queue);
+        m_queue.erase(m_queue.begin() + range.first, m_queue.begin() + range.first + range.second);
+    }
+
+    // dequeues all notifications from the queue
+    void dequeueElements(UniValue& result)
+    {
+        queueRange_t range = weakDequeueNotifications(result);
+        eraseRangeFromQueue(range);
+    }
+
+    bool elementsAvailable()
+    {
+        LOCK(m_cs_notification_queue);
+        return m_queue.size() > 0;
+    }
+
+    void registerType(NotificationType type)
+    {
+        if (type == NotificationType::Unknown)
+            return;
+
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.insert(type);
+    }
+
+    void unregisterType(NotificationType type)
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.erase(type);
+    }
+
+    void unregisterAllTypes()
+    {
+        LOCK(m_cs_notification_queue);
+        m_registered_notification_types.clear();
+    }
+
+    void addToQueue(NotificationEntry entry)
+    {
+        LOCK(m_cs_notification_queue);
+
+        size_t queueSize = m_queue.size();
+        if (queueSize > MAX_QUEUE_SIZE) {
+            m_queue.pop_front();
+            LogPrintf("RPC Notification limit has been reached, dropping oldest element\n");
```

Comment:
I'd be somwhat worried we'd fill someone's drive with debug log entries in case they forget to deregister a listener, here.